core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala

```diff
@@ -243,7 +243,16 @@ private[spark] object ReliableCheckpointRDD extends Logging {
       if (fs.exists(partitionerFilePath)) {
         val fileInputStream = fs.open(partitionerFilePath, bufferSize)
         val serializer = SparkEnv.get.serializer.newInstance()
-        val deserializeStream = serializer.deserializeStream(fileInputStream)
+        // make sure that the file is closed if error occurrs during deserialization
+        val deserializeStream =
```

Comment:
@srowen Probably some code will explain things better. Do you mean this:

```scala
        var deserializeStream : DeserializationStream = null

        val partitioner = Utils.tryWithSafeFinally[Partitioner] {
          deserializeStream = serializer.deserializeStream(fileInputStream)
          deserializeStream.readObject[Partitioner]
        } {
          deserializeStream.close()
        }
```
There are a few things I am concerning
1. The semantic has slightly changed (`deserializeStream` is now `var` instead of `val`)
2. I'm not sure is it always safe to close an partially initialized `deserializeStream`, in the case of the deserialization throwing exception. I'm pretty sure that if we close the `fileInputStream` first may cause the closing `deserializeSteam` throwing exception complaining that the input has been closed already. 

FYI, I found this problem by while running the test case "checkpointing partitioners" in which the `corruptPartitionerFile` flag is turned on in the `CheckpointSuite`.